test/test_reason.re

```diff
@@ -1093,6 +1150,126 @@ let testWithId = () => {
   });
 };
 
+let testPropMaybeNone = () =>
+  withContainer(c => {
+    act(() =>
+      React.Dom.render(<div className=?None />, Dom_html.element(c))
+    );
+    assert_equal(c##.innerHTML, Js.string("<div></div>"));
+  });
+
+let testPropMaybeSome = () =>
+  withContainer(c => {
+    act(() =>
+      React.Dom.render(
+        <div className=?{Some("foo")} />,
+        Dom_html.element(c),
+      )
+    );
+    assert_equal(c##.innerHTML, Js.string("<div class=\"foo\"></div>"));
+  });
+
+let testPropCustomString = () =>
+  withContainer(c => {
+    module Prop = {
+      let foo = Prop.string("foo");
+    };
+    act(() => React.Dom.render(<div foo="bar" />, Dom_html.element(c)));
+    assert_equal(c##.innerHTML, Js.string("<div foo=\"bar\"></div>"));
+  });
+
+let testPropCustomBool = () =>
+  withContainer(c => {
+    module Prop = {
+      let disabled = Prop.bool("disabled");
+    };
+    act(() => React.Dom.render(<div disabled=true />, Dom_html.element(c)));
+    assert_equal(c##.innerHTML, Js.string("<div disabled=\"\"></div>"));
+  });
+
+let testPropCustomInt = () =>
+  withContainer(c => {
+    module Prop = {
+      let foo = Prop.int("foo");
+    };
+    act(() => React.Dom.render(<div foo=42 />, Dom_html.element(c)));
+    assert_equal(c##.innerHTML, Js.string("<div foo=\"42\"></div>"));
+  });
+
+let testPropCustomFloat = () =>
+  withContainer(c => {
+    module Prop = {
+      let foo = Prop.float_("foo");
+    };
+    act(() => React.Dom.render(<div foo=42.5 />, Dom_html.element(c)));
+    assert_equal(c##.innerHTML, Js.string("<div foo=\"42.5\"></div>"));
+  });
+
+let testPropCustomAny = () =>
+  withContainer(c => {
+    module Prop = {
+      let foo: Js.t(Js.js_array(string)) => Prop.t = Prop.any("foo");
+    };
+    act(() =>
+      React.Dom.render(
+        <div foo={Js.array([|"bar", "baz"|])} />,
+        Dom_html.element(c),
+      )
+    );
+    assert_equal(c##.innerHTML, Js.string("<div foo=\"bar,baz\"></div>"));
+  });
+
+let testCustomElement = () =>
+  withContainer(c => {
+    let cool_element = h("cool-element");
+    let chill = h("chill");
+    module Prop = {
+      let coolness = Prop.string("coolness");
+      let data_out = Prop.bool("data-out");
+    };
+    act(() =>
+      React.Dom.render(
+        <cool_element coolness="max"> <chill data_out=true /> </cool_element>,
```

Comment:
Example of custom elements + jsx.

Reply:
Oh wow, React will already do that? I did not know!

Reply:
React will create DOM elements with whatever tag is passed to `createElement`. Here we pass `cool-element` through the `h` function:

https://github.com/ml-in-barcelona/jsoo-react/blob/d3924adb125a8a7fadb8c2535486127c9b05c1fa/test/test_reason.re#L1224

Because we have not defined these elements Right now we get a warning when running the tests:

```
Warning: The tag <chill> is unrecognized in this browser. If you meant to render a React component, start its name with an uppercase letter.
    in chill
    in cool-element
```

I am not sure I get it, which is the part you did not know about? :)

Reply:
Oh, I thought React was the one translating `cool_element` to `cool-element`, but it's the PPX. Yeah that's a bit of a shame with the warning, it ought to accept standard custom element names I think.

Reply:
I think it'd work if we had previously registered `cool-element` and `chill` as custom elements using `window.customElements.define`. I tried, but unfortunately jsoo won't parse `class Foo extends HTMLElement {...}`... so I guess we will have to live with this warning for now.

Reply:
Hmm, I just got this to work in #129. Didn't try Reason syntax though. I'm not sure how React would check the custom element rehister though, or if that's even possible.

Reply:
> Didn't try Reason syntax though. I'm not sure how React would check the custom element register though, or if that's even possible.

It does some hacks to check if the element has been registered apparently: https://github.com/facebook/react/blob/24dd07bd269590ee5024b7f0f1906887d256ea86/packages/react-dom/src/client/ReactDOMComponent.js#L447-L461

I don't understand the `if` condition though, as it uses `if (!isCustomComponentTag && ...`, but I would expect `if (isCustomComponentTag && ...`, without negation 🤔 

> Didn't try Reason syntax though.

Nothing should change, right? The Reason syntax conversion of JSX elements is now super straight forward.

Reply:
> I don't understand the `if` condition though, as it uses `if (!isCustomComponentTag && ...`, but I would expect `if (isCustomComponentTag && ...`, without negation thinking

I don't know what `isCustomComponent` means, but I think it's the second clause, `Object.prototype.toString.call(domElement) === '[object HTMLUnknownElement]'`. The trick seems to be to check after instantiating the dom element, whether it instantiated a known or unknown element.

> Nothing should change, right? The Reason syntax conversion of JSX elements is now super straight forward.

That's true! I guess it's alright then :)